File: DSA/Ishita-Nigam/Day74/Day74_Q1.java
//Maximum Product of Splitted Binary Tree
import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) {
        this.val = val;
    }
}

public class Day74_Q1 {
    public int maxProduct(TreeNode root) {
        final int kMod = 1_000_000_007;
        long ans = 0;
        List<Integer> allSums = new ArrayList<>();
        final long totalSum = treeSum(root, allSums);

        for (final long sum : allSums)
            ans = Math.max(ans, sum * (totalSum - sum));

        return (int) (ans % kMod);
    }

    private int treeSum(TreeNode root, List<Integer> allSums) {
        if (root == null)
            return 0;

        final int leftSum = treeSum(root.left, allSums);
        final int rightSum = treeSum(root.right, allSums);
        final int sum = root.val + leftSum + rightSum;
        allSums.add(sum);
        return sum;
    }

    public static void main(String[] args) {
        // Example usage:
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);

        Day74_Q1 solution = new Day74_Q1();
        System.out.println(solution.maxProduct(root));  // Output will depend on the tree structure
    }
}

File: DSA/Aarohi-Sharma/Day74_q1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
int totalSum(TreeNode* root){
        if(!root){return 0;}
        return root->val+totalSum(root->left)+totalSum(root->right);
    }
    pair<int,unsigned long long> solve(TreeNode* root,unsigned long long &sum){
        if(!root){return {0,0};}
        pair<unsigned long long,unsigned long long> left=solve(root->left,sum);
        pair<unsigned long long,unsigned long long> right=solve(root->right,sum);
        return {root->val+left.first+right.first,max(max((sum-left.first)*left.first,(sum-right.first)*right.first),max(left.second,right.second))};
    }
public:
    int maxProduct(TreeNode* root){
        unsigned long long sum=totalSum(root);
        pair<unsigned long long,unsigned long long> ans=solve(root,sum);
        int mod_ans=ans.second%(int(1e9+7));
        return mod_ans;
        
    }
};
